cli/cmd/check.go

```diff
@@ -161,14 +169,93 @@ func runChecks(w io.Writer, hc *healthcheck.HealthChecker) bool {
 			}
 		}
 
-		fmt.Fprintf(w, "%s %s\n", status, result.Description)
+		fmt.Fprintf(wout, "%s %s\n", status, result.Description)
 		if result.Err != nil {
-			fmt.Fprintf(w, "    %s\n", result.Err)
+			fmt.Fprintf(wout, "    %s\n", result.Err)
 			if result.HintAnchor != "" {
-				fmt.Fprintf(w, "    see %s%s for hints\n", healthcheck.HintBaseURL, result.HintAnchor)
+				fmt.Fprintf(wout, "    see %s%s for hints\n", healthcheck.HintBaseURL, result.HintAnchor)
+			}
+		}
+	}
+
+	success := hc.RunChecks(prettyPrintResults)
+	// this empty line separates final results from the checks list in the output
+	fmt.Fprintln(wout, "")
+
+	if !success {
+		fmt.Fprintf(wout, "Status check results are %s\n", failStatus)
+	} else {
+		fmt.Fprintf(wout, "Status check results are %s\n", okStatus)
+	}
+
+	return success
+}
+
+type checkCategory struct {
+	Name   string   `json:"categoryName"`
+	Checks []*check `json:"checks"`
+}
+
+type check struct {
+	Description string      `json:"description"`
+	Hint        string      `json:"hint"`
+	Result      checkResult `json:"result"`
+}
+
+type checkResult string
+
+const (
+	checkSuccess checkResult = "success"
+	checkWarn    checkResult = "warning"
+	checkErr     checkResult = "error"
+)
+
+func runChecksJSON(wout io.Writer, werr io.Writer, hc *healthcheck.HealthChecker) bool {
+	var lastCategory healthcheck.CategoryID
+	var outputJSON []*checkCategory
+	var currentCategory *checkCategory
+
+	collectJSONOutput := func(result *healthcheck.CheckResult) {
```

Comment:
Not 100% sure if this is equivalent to what we are trying to accomplish here, but here's a suggested implementation of `runChecksJSON()` without needing the new `checkCategory`, `check` and `checkResult` types:

```diff
diff --git a/cli/cmd/check.go b/cli/cmd/check.go
index a32b3556..13ca9f2c 100644
--- a/cli/cmd/check.go
+++ b/cli/cmd/check.go
@@ -211,42 +211,15 @@ const (
 )
 
 func runChecksJSON(wout io.Writer, werr io.Writer, hc *healthcheck.HealthChecker) bool {
-	var lastCategory healthcheck.CategoryID
-	var outputJSON []*checkCategory
-	var currentCategory *checkCategory
-
+	outputJSON := map[healthcheck.CategoryID][]*healthcheck.CheckResult{}
 	collectJSONOutput := func(result *healthcheck.CheckResult) {
-		if lastCategory != result.Category {
-			currentCategory = &checkCategory{
-				Name:   string(result.Category),
-				Checks: []*check{},
-			}
-			outputJSON = append(outputJSON, currentCategory)
-
-			lastCategory = result.Category
+		_, exists := outputJSON[result.Category]
+		if !exists {
+			outputJSON[result.Category] = []*healthcheck.CheckResult{}
 		}
 
 		if !result.Retry {
-			// ignore checks that are going to be retried, we want only final results
-			status := checkSuccess
-			if result.Err != nil {
-				status = checkErr
-				if result.Warning {
-					status = checkWarn
-				}
-			}
-
-			var hint string
-			if result.Err != nil && result.HintAnchor != "" {
-				hint = fmt.Sprintf("%s%s", healthcheck.HintBaseURL, result.HintAnchor)
-			}
-
-			currentCheck := &check{
-				Description: result.Description,
-				Result:      status,
-				Hint:        hint,
-			}
-			currentCategory.Checks = append(currentCategory.Checks, currentCheck)
+			outputJSON[result.Category] = append(outputJSON[result.Category], result)
 		}
 	}
```

Output:
```
{
  "kubernetes-api": [
    {
      "Category": "kubernetes-api",
      "Description": "can initialize the client",
      "HintAnchor": "k8s-api",
      "Retry": false,
      "Warning": false,
      "Err": null
    },
    {
      "Category": "kubernetes-api",
      "Description": "can query the Kubernetes API",
      "HintAnchor": "k8s-api",
      "Retry": false,
      "Warning": false,
      "Err": null
    }
  ],
  "kubernetes-version": [
    {
      "Category": "kubernetes-version",
      "Description": "is running the minimum Kubernetes API version",
      "HintAnchor": "k8s-version",
      "Retry": false,
      "Warning": false,
      "Err": null
    },
    {
      "Category": "kubernetes-version",
      "Description": "is running the minimum kubectl version",
      "HintAnchor": "kubectl-version",
      "Retry": false,
      "Warning": false,
      "Err": null
    }
  ],
  "linkerd-existence": [
    {
      "Category": "linkerd-existence",
      "Description": "control plane namespace exists",
      "HintAnchor": "l5d-existence-ns",
      "Retry": false,
      "Warning": false,
      "Err": {}
    }
  ]
}
```

Reply:
nice! I removed two of those variables. I still kept the custom structs so the output is a bit different :)